### PR TITLE
[DIA-2198] Manipulate CCPA consent string based on applies

### PR DESCRIPTION
**Pull request by bohdan-go-wombat**

### CODE REVIEW

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt [hunk 1]
```diff
@@ -540,31 +542,50 @@ private class CampaignManagerImpl(
...
-                sampleRate?.let { i ->
-                    if (i != dataStorage.ccpaSamplingValue) {
-                        dataStorage.ccpaSamplingValue = i
-                        dataStorage.ccpaSamplingResult = null
-                    }
+        get() = dataStorage.gdprConsentStatus != null || dataStorage.uspstring.isNullOrEmpty().not()
+
+    override fun handleMetaDataResponse(response: MetaDataResp?) {
+        // update meta data response in the data storage
+        metaDataResp = response
```
Comment (andresilveirah) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt:549]: @bohdan-go-wombat maybe this comment doesn't belong to this PR, but do we need to store the responses? 

If you think it makes sense, we could remove `metaDataResp` from the local storage already.

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt [hunk 2]
```diff
@@ -540,31 +542,50 @@ private class CampaignManagerImpl(
...
+
+        if (response == null) return
+
+        // handle ccpa
+        response.ccpa?.apply {
+            applies?.let { i ->
```
Comment (andresilveirah) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt:555]: @bohdan-go-wombat could you rename `i` to something more intuitive in the lines below? Otherwise I don't see the improvement using `i` over `it`.

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt [hunk 3]
```diff
@@ -540,31 +542,50 @@ private class CampaignManagerImpl(
...
+                if (i != dataStorage.ccpaSamplingValue) {
+                    dataStorage.ccpaSamplingValue = i
+                    dataStorage.ccpaSamplingResult = null
+                }
+            }
+        }
+
+        // handle gdpr
+        response.gdpr?.apply {
+            applies?.let { i -> dataStorage.gdprApplies = i }
```
Comment (andresilveirah) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt:573]: I thought we'd get rid of `dataStorage.gdprApplies` and deal only with the consent object `.applies` attribute.

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorage.kt [hunk 4]
```diff
@@ -35,6 +35,7 @@ internal interface DataStorage : DataStorageGdpr, DataStorageCcpa {
 
     var consentStatusResponse: String?
     var gdprConsentStatus: String?
+    val ccpaApplies: Boolean
```
Comment (andresilveirah) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorage.kt:38]: Why did you decide to move `ccpaApplies` from `DataStorageCcpa` class here?

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageImpl.kt [hunk 5]
```diff
@@ -79,6 +85,18 @@ private class DataStorageImpl(
                 .apply()
         }
 
+    /**
+     * Parameter that accesses applies value from ccpaConsentStatus directly
+     *
+     * By default - false (if there are no ccpaConsentStatus in data storage or applies value is null)
+     */
+    override val ccpaApplies: Boolean
```
Comment (andresilveirah) [inline cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageImpl.kt:93]: @bohdan-go-wombat what about the `set`? If I understood it right, we access `ccpaApplies` by reaching out to the ccpa consent object, but can still set its value independently. Is that right?

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorage.kt (hunk 4 above)
Comment (bohdan-go-wombat) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorage.kt:38]: @andresilveirah because it is a "mirroring" parameter, and we decided to actually access it via `ccpaConsentStatus` from the DataStorageCcpa

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt (hunk 2 above)
Comment (bohdan-go-wombat) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt:555]: @andresilveirah agree

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt (hunk 1 above)
Comment (bohdan-go-wombat) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt:549]: @andresilveirah I think comments like that could actually help a lot with a readability of the code for the future refferences

regarding `metaDataResp`, I will take a look, but as I already mentioned, I strongly not recommend do a lot of refactoring in the PR, that is responsible for a speciffic fix/feature

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt (hunk 3 above)
Comment (bohdan-go-wombat) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt:573]: @andresilveirah since this is a PR to fix CCPA related issue, I decided to refactor CCPA related code only, and for the GDPR code I can create a ticket to do it separatelly, since:
1. it will help us to merge this task asap
2. it will prevent us from making mistakes in the part of the code, that is not related to the issue we are trying to fix
3. it will prevent us from breaking the purpose of the pull request (which is to fix a problem with uspstring)

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageImpl.kt (hunk 5 above)
Comment (bohdan-go-wombat) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageImpl.kt:93]: when I refactored I made `ccpaApplies` `val`, since we no longer need to update the `mirroring` value, and we will just update the local `ccpaConsentStatus` object (the one that is stored in the data store and being used in our app)

Comment (bohdan-go-wombat) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/data/local/DataStorageImpl.kt:93]: getter is needed to have the easier access to the `ccpaApplies`, for example rather then calling something like:
`dataStorage.ccpaConsentStatus?.let { verifiedCcpaConsentStatusString ->
            check { JsonConverter.converter.decodeFromString<CcpaCS>(verifiedCcpaConsentStatusString) }
                .getOrNull()
                ?.applies`
we will call:
`dataStorage.ccpaApplies`

File: cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt (hunk 1 above)
Comment (andresilveirah) [reply cmplibrary/src/main/java/com/sourcepoint/cmplibrary/campaign/CampaignManager.kt:549]: @bohdan-go-wombat my bad, the comment I was referring to was my GH comment, not the comment on the code 😅 

Regarding not mixing all the refactoring in the same PR, agree with you.

**Review by Nevazhnovu — APPROVED** [review]

**Review by andresilveirah — APPROVED** [review]
